runtime/src/message_processor.rs

```diff
@@ -104,9 +102,13 @@ impl MessageProcessor {
             }
 
             invoke_context.set_instruction_index(instruction_index);
+            let mut pre_remaining_units = 0u64;
             let result = invoke_context
                 .push(message, instruction, program_indices, None)
                 .and_then(|_| {
+                    // read the `pre` value after potential meter reset, right before processing
+                    pre_remaining_units =
+                        invoke_context.get_compute_meter().borrow().get_remaining();
```

Comment:
We could read the post-units in this `and_then` and do the calc here as well, might make it a bit tidier.
```
units consumed = 0
push and then
read pre units
process instruction
read post units
units-consumed = post - pre
...
accumulate (program_id, time, units_consumed)
```

Reply:
good idea! moved timer and unit measure into and_then, closer to process instruction.